### PR TITLE
Make Test Functions flexible for MultiObjective Optimizers

### PR DESCRIPTION
**Pull request by jonpsy**
Following the discussion of #272.  It was decided that adding bounds to the problem function is convenient (**but not necessary**) and is mostly restricted for testing purposes.

Once this PR gets merged we can have a series of followup (or even continue this?) PR to complete the task.

**LOGS**
- [x] ackley_function.hpp: Molga, Marcin, and Czesław Smutnicki. "Test functions for optimization needs." Test functions for optimization needs 101 (2005): 48.  
- [ ] easom_function.hpp                   
- [ ] himmelblau_function.hpp          
- [ ] rastrigin_function.hpp        
- [x] beale_function.hpp : Wikipedia
- [ ] eggholder_function.hpp
- [ ] holder_table_function.hpp
- [ ] rosenbrock_function.hpp
- [x] booth_function.hpp  : Wikipedia
- [ ] fonseca_fleming_function.hpp         
- [ ] levy_function_n13.hpp             
- [ ] rosenbrock_wood_function.hpp 
- [ ] sparse_test_function.hpp
- [ ] bukin_function.hpp         
- [ ] fw_test_function.hpp
- [x] schaffer_function_n1.hpp: "Multiple Objective Optimization with Vector Evaluated Genetic Algorithms" (Wikipedia suggest -10 to -10^5 for lowerBound is OK so i went with -10^3).
- [ ] sphere_function.hpp
- [ ] colville_function.hpp       
- [ ] generalized_rosenbrock_function.hpp  
- [ ] matyas_function.hpp               
- [ ] schaffer_function_n2.hpp      
- [ ] styblinski_tang_function.hpp
- [ ] cross_in_tray_function.hpp 
- [ ] goldstein_price_function.hpp         
- [ ] mc_cormick_function.hpp           
- [ ] schaffer_function_n4.hpp      
- [ ] three_hump_camel_function.hpp
- [ ] drop_wave_function.hpp      
- [ ] schwefel_function.hpp       
- [ ] wood_function.hpp

### CODE REVIEW

File: include/ensmallen_bits/problems/ackley_function.hpp [hunk 1]
```diff
@@ -112,10 +112,10 @@ class AckleyFunction
   //! Modify the value used for numerical stability.
   double& Epsilon() { return epsilon; }
 
-  // Note: GetInitialPoint(), GetFinalPoint(), and GetFinalObjective() are not
-  // required for using ensmallen to optimize this function!  They are
-  // specifically used as a convenience just for ensmallen's testing
-  // infrastructure.
+  // Note: GetInitialPoint(), GetFinalPoint(), and GetFinalObjective(),
+  // GetVariablesBounds() are not required for using ensmallen to
```
Comment (rcurtin) [inline include/ensmallen_bits/problems/ackley_function.hpp:116]:
```suggestion
  // Note: GetInitialPoint(), GetFinalPoint(), GetFinalObjective(), and
  // GetVariablesBounds() are not required for using ensmallen to
```
Just a quick grammar fix. :)

File: include/ensmallen_bits/problems/ackley_function.hpp [hunk 2]
```diff
@@ -128,6 +128,15 @@ class AckleyFunction
   //! Get the final objective.
   double GetFinalObjective() const { return 0.0; }
 
+  //! Get the bounds of variable space.
+  std::tuple<arma::vec, arma::vec> GetVariablesBounds() const
```
Comment (rcurtin) [inline include/ensmallen_bits/problems/ackley_function.hpp:132]: Maybe `GetVariableBounds()`?

**Review by rcurtin — COMMENTED** [review]
Looks good to me---want to take it a step further and make something like `FunctionTest<>` (in `tests/test_function_tools.hpp`) for multi-objective optimizers?

**Comment by jonpsy** [comment]
@rcurtin , I'm game.

**Comment by rcurtin** [comment]
Sure, definitely; I haven't kept up with the NSGA-II PR, but whenever that's ready and you're ready to come back to this, I'll try to review it. :+1:  Ideally I think it would be better to do he refactoring to add bounds and also modify the tests all in one PR---I don't see a need to spread it over multiple PRs.

**Comment by jonpsy** [comment]
Okay, I've begun working on this. Check the logs for updates. I can 2 functions daily and this will be done within a week and a half I guess. That'll give sufficient time for reviews as well.  @rcurtin  I've also begun thinking about the test_function_tools you've talked about, expect some updates soon.

**Pull request closed without merging**